compiler/AST/symbol.cpp

```diff
@@ -2994,22 +2937,158 @@ void LabelSymbol::accept(AstVisitor* visitor) {
 // with C escapes - that is newline is 2 chars \ n
 // so this function expects a string that could be in "" in C
 VarSymbol *new_StringSymbol(const char *str) {
+  SET_LINENO(stringLiteralModule);
+
+  // Hash the string and return an existing symbol if found.
+  // Aka. uniqify all string literals
+  Immediate imm;
+  imm.const_kind = CONST_KIND_STRING;
+  imm.string_kind = STRING_KIND_STRING;
+  imm.v_string = astr(str);
+  VarSymbol *s = stringLiteralsHash.get(&imm);
+  if (s) {
+    return s;
+  }
+
+  if (resolved) {
+    INT_FATAL("new_StringSymbol called after function resolution.");
+  }
+
+  // String (as record) literals are inserted from the very beginning on the
+  // parser all the way through resolution (postFold). Since resolution happens
+  // after normalization we need to insert everything in normalized form. We
+  // also need to disable parts of normalize from running on literals inserted
+  // at parse time.
+
+  VarSymbol* cptrTemp = newTemp("call_tmp");
+  cptrTemp->addFlag(FLAG_TYPE_VARIABLE);
+  CallExpr *cptrCall = new CallExpr(PRIM_MOVE,
+      cptrTemp,
+      new CallExpr("_type_construct_c_ptr", new SymExpr(dtUInt[INT_SIZE_8]->symbol)));
+
+  VarSymbol* castTemp = newTemp("call_tmp");
+  CallExpr *castCall = new CallExpr(PRIM_MOVE,
+      castTemp,
+      new CallExpr("_cast", cptrTemp, new_CStringSymbol(str)));
+
+  int strLength = unescapeString(str).length();
+
+  CallExpr *ctor = new CallExpr("_construct_string",
+      castTemp,
+      new_IntSymbol(strLength),   // length
+      new_IntSymbol(strLength ? strLength+1 : 0), // size, empty string needs 0
+      gFalse);                    // owned = false
+  ctor->insertAtTail(gFalse);     // needToCopy = false
+
+  s = new VarSymbol(astr("_str_literal_", istr(literal_id++)), dtString);
+  s->addFlag(FLAG_NO_AUTO_DESTROY);
+  s->addFlag(FLAG_CONST);
+  DefExpr* stringLitDef = new DefExpr(s);
+  // DefExpr(s) always goes into the module scope to make it a global
+  stringLiteralModule->block->insertAtTail(stringLitDef);
+
+  CallExpr* ctorCall = new CallExpr(PRIM_MOVE, new SymExpr(s), ctor);
+
+  if (stringLiteralModule->initFn) {
+    stringLiteralModule->initFn->insertBeforeReturn(new DefExpr(cptrTemp));
+    stringLiteralModule->initFn->insertBeforeReturn(cptrCall);
+    stringLiteralModule->initFn->insertBeforeReturn(new DefExpr(castTemp));
+    stringLiteralModule->initFn->insertBeforeReturn(castCall);
+    stringLiteralModule->initFn->insertBeforeReturn(ctorCall);
+
+  } else {
+    // Modules dont have an initFn until normalization happens
+    // These Exprs will be moved into the initFn when it is created.
+    stringLiteralModule->block->insertAtTail(new DefExpr(cptrTemp));
+    stringLiteralModule->block->insertAtTail(cptrCall);
+    stringLiteralModule->block->insertAtTail(new DefExpr(castTemp));
+    stringLiteralModule->block->insertAtTail(castCall);
+    stringLiteralModule->block->insertAtTail(ctorCall);
+  }
```

Comment:
Looks like this function has to support being called at various points in compilation, which complicates it significantly. I *think* it would be simpler if we were creating local variable string records out of C strings. I'd consider renaming this method, and only supporting new_CStringSymbol in most places. Then, the renamed method would create a local variable string record (I don't think it needs to be global, even though that's how it used to work).